bbq/compiler/peephole_pass.go

```diff
@@ -0,0 +1,202 @@
+		case opcode.InstructionJumpIfFalse:
+			optimized[jump] = opcode.InstructionJumpIfFalse{Target: uint16(newJumpTarget)}
+		case opcode.InstructionJumpIfTrue:
+			optimized[jump] = opcode.InstructionJumpIfTrue{Target: uint16(newJumpTarget)}
+		case opcode.InstructionJumpIfNil:
+			optimized[jump] = opcode.InstructionJumpIfNil{Target: uint16(newJumpTarget)}
+		}
+	}
+}
+
+func (o *PeepholeInstructionOptimizer[E, T]) OptimizeInstructions(instructions []opcode.Instruction) []opcode.Instruction {
+	optimized := make([]opcode.Instruction, 0, len(instructions))
+
+	for i := 0; i < len(instructions); i++ {
+		currentInstruction := instructions[i]
+		candidates := o.PatternsByOpcode[currentInstruction.Opcode()]
+
+		var matched bool
+		// check candidates for pattern match for each instruction
+		for _, candidate := range candidates {
+			candidateOpcodes := candidate.Opcodes
+			window := instructions[i : i+len(candidateOpcodes)]
+
+			if candidate.Match(window) {
+				replacement := candidate.Replacement(window, o.compiler)
+				optimized = append(optimized, replacement...)
+
+				o.bytecodeShifts = append(o.bytecodeShifts, BytecodeShiftPair{
+					Offset: i,
+					// e.g. replace 2 with 1 -> shift is -1
+					Shift: len(replacement) - len(candidateOpcodes),
+				})
+
+				i += len(candidateOpcodes) - 1
+				matched = true
+				break
+			}
+		}
+		if !matched {
+			optimized = append(optimized, currentInstruction)
+			currentOpcode := currentInstruction.Opcode()
+
+			// if instruction is a jump, add it to the jumps list
+			// its important we do it in the optimized index, not the original index
+			// because its hard to calculate the optimized index from the original index later
+			if currentOpcode == opcode.Jump ||
+				currentOpcode == opcode.JumpIfFalse ||
+				currentOpcode == opcode.JumpIfTrue ||
+				currentOpcode == opcode.JumpIfNil {
+				o.jumps = append(o.jumps, len(optimized)-1)
+			}
+		}
+	}
+
+	// Sort jumps by their target offsets
+	sort.Slice(o.jumps, func(a, b int) bool {
+		var jumpATarget, jumpBTarget uint16
+		switch ins := optimized[o.jumps[a]].(type) {
+		case opcode.InstructionJump:
+			jumpATarget = ins.Target
+		case opcode.InstructionJumpIfFalse:
+			jumpATarget = ins.Target
+		case opcode.InstructionJumpIfTrue:
+			jumpATarget = ins.Target
+		case opcode.InstructionJumpIfNil:
+			jumpATarget = ins.Target
+		}
+
+		switch ins := optimized[o.jumps[b]].(type) {
+		case opcode.InstructionJump:
+			jumpBTarget = ins.Target
+		case opcode.InstructionJumpIfFalse:
+			jumpBTarget = ins.Target
+		case opcode.InstructionJumpIfTrue:
+			jumpBTarget = ins.Target
+		case opcode.InstructionJumpIfNil:
+			jumpBTarget = ins.Target
+		}
+		// ascending order
+		return jumpATarget < jumpBTarget
+	})
+
+	o.patchJumps(optimized)
+
+	return optimized
+}
+
+func (o *PeepholeInstructionOptimizer[E, T]) Optimize(code []E) []E {
+	if instructions, ok := any(code).([]opcode.Instruction); ok {
```

Comment:
If `PeepholeInstructionOptimizer` only works for `E == opcode.Instruction`, then the `E` type parameter should be removed. Happy to help with that